File: 0020-valid-parentheses/solution.c
bool isValid(char *s) {
    int MAX_SIZE = 10000;
    char stack[MAX_SIZE];
    int top = -1;
    char brackets_map[128];

    brackets_map[')'] = '(';
    brackets_map['}'] = '{';
    brackets_map[']'] = '[';

    for (int i = 0; s[i] != '\0'; i++) {
        if (s[i] == '(' || s[i] == '{' || s[i] == '[') {
            stack[++top] = s[i];
        } else if (s[i] == ')' || s[i] == '}' || s[i] == ']') {
            if (top == -1 || brackets_map[s[i]] != stack[top]) {
                return false;
            }
            top--;
        } else {
            return false;
        }
    }

    return top == -1;
}

File: 0020-valid-parentheses/solution.rs
impl Solution {
    pub fn is_valid(s: String) -> bool {
        let mut stack: Vec<char> = Vec::new();

        for i in s.chars() {
            if i == '(' || i == '[' || i == '{' {
                stack.push(i);
            } else {
                if let Some(val) = stack.pop() {
                    if i == ']' && val != '[' {
                        return false;
                    }
                    if i == ')' && val != '(' {
                        return false; 
                    }
                    if i == '}' && val != '{' {
                        return false; 
                    }
                } else {
                    return false;
                }
            }
        }

        stack.len() == 0
    }
}
